### PR TITLE
Add support for JOINs with ON clause

### PR DESCRIPTION
**Pull request by ianmcodes**
These changes add support for different JOINs in SELECT in a manner similar to Perl SQL::Abstract. Added to the select test to test the new feature. Also improved syntax to reflect JavaScript block scoping of variables, and generally accepted best practices in creating arrays and accessing object properties.

Below is an example of using the new join feature and the sql generated.

``` javascript
var SqlG = require('sql-generator');
var sqlg = new SqlG();

var tables = [
    "T1",
    {
        "INNER_JOIN": { "T1.this": "T2.that" } 
    }, 
    "T2", 
    { 
        "INNER_JOIN": { "T2.some": "T3.other", "T2.i": { "!=": "T3.d" } } 
    }, 
    "T3"
];
var stmt = sqlg.select(tables, "*");
// stmt.sql => 'SELECT * FROM T1 INNER JOIN T2 ON T1.this = T2.that INNER JOIN T3 ON T2.some = T3.other AND T2.i != T3.d'
```

### CODE REVIEW

**Comment by shot056** [comment]
Thank you for PR.

I have some question.
please reply to these.

File: lib/sql-generator/index.js [hunk 1]
```diff
@@ -27,41 +28,41 @@ var SqlGenerator = function( type ) {
...
-            sqls.push( 'OFFSET ' + opt['offset'] );
+        if( opt.limit && typeof opt.limit === 'number' )
+            sqls.push( 'LIMIT ' + opt.limit );
+        if( opt.offset && typeof opt.offset === 'number' )
+            sqls.push( 'OFFSET ' + opt.offset );
         return sqls.join(' ');
     };
     self._gen_where = function( tmp_where, values ) {
-        var w;
+        var w, ws;
```
Comment (shot056) [inline lib/sql-generator/index.js:54]: why did you move the declaration on the outside of block ?

**Comment by ianmcodes** [comment]
JavaScript variables (declared using `var`) are function scoped, not block scoped as you would see in other languages. Here are a couple of pages that explain variable scope in JavaScript:
- http://javascriptissexy.com/javascript-variable-scope-and-hoisting-explained/
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var

Also that variable was being declared twice in that function.

**Comment by shot056** [comment]
OK. I understood.

**Comment by ianmcodes** [comment]
Thanks :smiley:
